include/boost/optional/detail/optional_utility.hpp

```diff
@@ -0,0 +1,38 @@
+// Copyright (C) 2024 Andrzej Krzemienski.
+//
+// Use, modification, and distribution is subject to the Boost Software
+// License, Version 1.0. (See accompanying file LICENSE_1_0.txt or copy at
+// http://www.boost.org/LICENSE_1_0.txt)
+//
+// See http://www.boost.org/libs/optional for documentation.
+//
+// You are welcome to contact the author at:
+//  akrzemi1@gmail.com
+
+#ifndef BOOST_OPTIONAL_OPTIONAL_DETAIL_OPTIONAL_UTILITY_HPP
+#define BOOST_OPTIONAL_OPTIONAL_DETAIL_OPTIONAL_UTILITY_HPP
+
+namespace boost {
+namespace optional_detail {
+
+// Workaround: forward and move aren't constexpr in C++11
+template <class T> inline constexpr T&& forward(typename boost::remove_reference<T>::type& t) noexcept
+{
+  return static_cast<T&&>(t);
+}
+
+template <class T> inline constexpr T&& forward(typename boost::remove_reference<T>::type&& t) noexcept
+{
+  BOOST_STATIC_ASSERT_MSG(!boost::is_lvalue_reference<T>::value, "Can not forward an rvalue as an lvalue.");
```

Comment:
Ok, so the C++11 (https://www.open-std.org/jtc1/sc22/wg21/docs/papers/2012/n3337.pdf) allows:
— null statements,
— static_assert-declarations
— typedef declarations and alias-declarations that do not define classes or enumerations,
— using-declarations,
— using-directives,
— and exactly one return statement;

I am not sure if `BOOST_STATIC_ASSERT_MSG` fits in any of these groups.

Reply:
Ok, maybe just guard the assertion with `BOOST_NO_CXX11_STATIC_ASSERT`.

Reply:
I am not sure of where to look in that standard paper, or if I'd even understand its meaning, so I will trust you.

>Ok, maybe just guard the assertion with BOOST_NO_CXX11_STATIC_ASSERT.

The BOOST_STATIC_ASSERT_MACRO is implemented as such; 
```c++
#ifndef BOOST_NO_CXX11_STATIC_ASSERT
#  ifndef BOOST_NO_CXX11_VARIADIC_MACROS
#     define BOOST_STATIC_ASSERT_MSG( ... ) static_assert(__VA_ARGS__)
#  else
#     define BOOST_STATIC_ASSERT_MSG( B, Msg ) static_assert( B, Msg )
#  endif
#else
#     define BOOST_STATIC_ASSERT_MSG( B, Msg ) BOOST_STATIC_ASSERT( B )
#endif
```
Do we still need to guart it with `BOOST_NO_CXX11_STATIC_ASSERT` ?

Reply:
It is in document https://wg21.link/std11, clause 7.1.5 ([dcl.constexpr]), paragraph 3.
I may be too cautious, bt just in case there are compilers that support rvalue references and variadic templates but no `static_assert`, do guard the static assert with `BOOST_NO_CXX11_STATIC_ASSERT`. In that case. THe normal `static_assert` will be fine, so:

```c++
{
  #ifndef BOOST_NO_CXX11_STATIC_ASSERT
  static_assert(!boost::is_lvalue_reference<T>::value, "Can not forward an rvalue as an lvalue.");
  #endif
  return static_cast<T&&>(t);
```